ethereum/contracts/bridge/token/TokenImplementation.sol

```diff
@@ -174,4 +209,110 @@ contract TokenImplementation is TokenState, Context {
 
         _;
     }
+
+    /**
+     * @dev Returns the domain separator for the current chain.
+     */
+    function _domainSeparatorV4() internal view returns (bytes32) {
+        if (address(this) == _state.cachedThis && block.chainid == _state.cachedChainId) {
+            return _state.cachedDomainSeparator;
+        } else {
+            return _buildNativeDomainSeparator(
+                _hashedDomainType(),
+                _hashedTokenChain(),
+                _hashedNativeContract(),
+                _hashedDomainVersion()
+            );
+        }
+    }
+
+    function _buildNativeDomainSeparator(
+        bytes32 typeHash,
+        bytes32 tokenChainHash,
+        bytes32 nativeContractHash,
+        bytes32 versionHash
+    ) internal view returns (bytes32) {
+        return keccak256(abi.encode(typeHash, tokenChainHash, nativeContractHash, versionHash, block.chainid, address(this)));
+    }
+
+    /**
+     * @dev Given an already https://eips.ethereum.org/EIPS/eip-712#definition-of-hashstruct[hashed struct], this
+     * function returns the hash of the fully encoded EIP712 message for this domain.
+     *
+     * This hash can be used together with {ECDSA-recover} to obtain the signer of a message. For example:
+     *
+     * ```solidity
+     * bytes32 digest = _hashTypedDataV4(keccak256(abi.encode(
+     *     keccak256("Mail(address to,string contents)"),
+     *     mailTo,
+     *     keccak256(bytes(mailContents))
+     * )));
+     * address signer = ECDSA.recover(digest, signature);
+     * ```
+     */
+    function _hashTypedDataV4(bytes32 structHash) internal view returns (bytes32) {
+        return ECDSA.toTypedDataHash(_domainSeparatorV4(), structHash);
+    }
+
+    /**
+     * @dev See {IERC20Permit-permit}.
+     */
+    function permit(
+        address owner_,
+        address spender_,
+        uint256 value_,
+        uint256 deadline_,
+        uint8 v_,
+        bytes32 r_,
+        bytes32 s_
+    ) public {
+        // for those tokens that have been initialized before permit, we need to set
+        // the permit state variables
+        if (!permitInitialized()) {
```

Comment:
I'd move this check into `_initializePermitState`. since it's an internal call it's just an extra jump, which is pretty cheap, but then `_initializePermitState` is nice and self-contained and covered on all entry points

Reply:
If I do that, I'd like to name the method `_initializePermitStateIfNeeded`. Is that too much?

Reply:
I agree w/ this change, and I think this function name is fine.

Reply:
Fixed. How does it look?